### ISSUE TITLE
vmware.py:find_vm_by_id() fails when a VM is deleted from the folder

### ISSUE THREAD

**Issue by karniemi** [ansible/ansible#37254]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

vmware_guest

##### ANSIBLE VERSION

```
2.4
```

##### SUMMARY

vmware.py:find_vm_by_id() fails when a VM is deleted from the folder during the search. The relevant piece of code:
* https://github.com/ansible/ansible/blob/5e827f4cc437bf3529da7fb0ff66875570c14c39/lib/ansible/module_utils/vmware.py#L215-L228

Technically, when looping  for f_obj.childEntity, a VM may get deleted from the folder during the looping. The piece of code referenced above does not catch the resulting exception and makes the search fail. The exception has the msg " 'The object has already been deleted or has not been completely created'". Note! it's not about deleting the VM which is being searched, but any of the other VMs in that same folder. Of course the bug appears only sporadically.

A similar issue has been faced in vmware/pyvmomi#445

##### STEPS TO REPRODUCE

```yaml

```

##### EXPECTED RESULTS
The find does not fail.

##### ACTUAL RESULTS

```

```

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/modules/cloud/vmware/vmware_guest.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/vmware/vmware_guest.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @Akasurde @akasurde @dagwieers @dav1x @jctanner @nerzhul @pdellaert
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @warthog9
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by wojciechkopciewicz** [comment]
We are also seeing this issue. Is anyone looking at this?

**Comment by Akasurde** [comment]
@wojciechkopciewicz @karniemi  Could you please provide playbook which can be used to reproduce this issue ?

**Comment by wojciechkopciewicz** [comment]
@Akasurde 
We see this issue when working on our custom module that is yet not ready to publish.
But the following playbook should do the job:

```
- name: Power vm off by name
  vmware_guest:
    hostname: 192.168.1.209
    username: administrator@vsphere.local
    password: vmware
    name: vmname
    datacenter: my_datacenter
    state: poweredoff
  delegate_to: localhost
```

Just run the playbook in a loop via vcenter and add/remove vms in parallel.

The fastest way to reproduce this was to add a loop in the module with the following code:
```
for i in range(1, 100):
    all_vms = get_all_objs(content, [vim.VirtualMachine])
```
I proposed a fix for this - see https://github.com/ansible/ansible/pull/40527.

**Comment by Akasurde** [comment]
resolved_by_pr #40527

**Comment by karniemi** [comment]
@Akasurde @wojciechkopciewicz Hi, not sure if I'm reading the code correctly in PR #40527...but to me it seems to fix a similar issue in another place. As far as I can see,  it does not address the bug in the piece of code that I pointed originally … in the description of this issue… in vmware.py:find_vm_by_id()-function. What do you think?

**Comment by wojciechkopciewicz** [comment]
@karniemi 
You are right, I concentrated too much on find_vm_by_name which is causing issue we are seeing. Root cause seems to be exactly the same here and similar fix should be done here. 
I wouldn't like to fix it in this PR for two reasons:
- it would be great if the fix for get_all_objs is included in 2.6
- I would need to reproduce the issue to make sure the fix works and does not break anything - it will take some time as so far I did not use find_vm_by_path so far.
@Akasurde should I raise a separate issue for find_vm_by_name?

**Comment by karniemi** [comment]
For us, the problem happened early in vmware_guest module execution. The code path was from this line:
https://github.com/ansible/ansible/blob/d14467b0294296b9a2b85e0ad150315c2d5a2491/lib/ansible/modules/cloud/vmware/vmware_guest.py#L1468
...with the VMs "name" as search criteria... up to the  vmware.py:find_vm_by_id(), which still has that bug in devel. Only, I'm not sure if find_vm_by_id() is used for searches with "name" anymore? Maybe this commit: https://github.com/ansible/ansible/commit/87f2893481f127a39f1c9ccd4ba7a5863d4f9d1b has already changed how the VMs are searched for with name? I hope that this new get_vm()-method and all the sub-methods does not blow up if one of the VMs is deleted in the middle of the search?

More generically: any piece of code that scans the objects in vcenter, should not blow up if some of the objects get deleted during the scan? Do the vmware modules / utils have a lot of such scanning…?

**Comment by wojciechkopciewicz** [comment]
@karniemi @Akasurde 
I managed to reproduce this with the following playbook:
```
- hosts: localhost
  gather_facts: False
  tasks:
    - vmware_vm_shell:
        username: vcenter_user
        password: password
        hostname: xx.xx.xx.xx
        folder: my_folder
        validate_certs: False
        vm_id_type: inventory_path
        vm_id: non_existent_vm
        vm_shell: echo
        vm_shell_args: not_important
        vm_shell_cwd: /tmp
        vm_username: user
        vm_password: password
```
Fix is on the way - verifying it right now

**Comment by wojciechkopciewicz** [comment]
@karniemi @Akasurde 
Looking at the code, the following modules/functions are potentially having the same problem (I may be wrong here - did not do any deep analysis):

vmware_vm_facts -> get_all_virtual_machines
vmware_resource_pool_facts -> gather_rp_facts
vmware_guest -> multiple places
vmware_drs_rule_facts -> gather_drs_rule_facts
vmware_datastore_maintenancemode -> potentially affected
vmware -> get_all_hosts_by_cluster, get_all_host_objs, find_obj, find_cluster_by_name_datacenter (why does find_obj not use find_all_objs?)
vmware_host -> check_host_state

Of course there can be more - as you wrote, issue will be seen whenever object is removed during collection iteration.

In my opinion vmware.get_all_objs and vmware.find_vm_by_id are most important as this is common code, used by different modules and at some point one needs to find vm by name to determine its uuid.
Maybe a separate issue should be open not to forget about other places.

**Comment by ansibot** [comment]
cc @ckotte
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @lparkes
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @Tomorrow9
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @goneri
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @pgbidkar
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by karniemi** [comment]
Any new on this? Within this timeframe there might have been some refactoring on the relevant pieces of code? Earlier @wojciechkopciewicz analysed that there's propabably 10s(?) of places that would need the fix -maybe after some proper refactoring only one place would need a fix (...hopeful thinking :-) )? After all, it makes no sense to have any piece of logic copy-pasted all around?

**Comment by ulysses-31** [comment]
I am also regularly seeing the Virtual Machine "has already been deleted or has not been completely created" error using the  Ansible 2.8.4 vmware_vm_facts module to enumerate our vSphere environments. Any news on this bug?

**Comment by ansibot** [comment]
Thank you very much for your interest in Ansible. Ansible has migrated much of the content into separate repositories to allow for more rapid, independent development. We are closing this issue/PR because this content has been moved to one or more collection repositories.

* lib/ansible/modules/cloud/vmware/vmware_guest.py -> https://galaxy.ansible.com/community/vmware

For further information, please see:
https://github.com/ansible/ansibullbot/blob/master/docs/collection_migration.md
